keps/sig-node/0036-20190118-ForceEviction.md

```diff
@@ -0,0 +1,170 @@
+
+to delete the Pods from API server without waiting for kubelet's acknoledgement.
+Volumes can also be safely detached since mounts no longer exist.
+
+## Non-Goals
+
+This proposal is not trying to address the recovery issue for all types of failures. For example for network
+partition issue, because this is no source of truth about whether Pods are running or not, it is not possible
+for controller decide what recovery actions should take based on the node status.
+
+## Proposal
+
+Currently, Kubernetes already supports four types of taint effects: 
+- NoSchedule: Pods that do not tolerate this taint are not scheduled on the node.
+- PreferNoSchedule: Kubernetes avoids scheduling Pods that do not tolerate this taint onto the node.
+- NoExecute: Pod is evicted from the node if it is already running on the node, and is not scheduled
+onto the node if it is not yet running on the node.
+
+We propose to add a new effect, TaintEffectForceEviction. For a Pod that cannot tolerant this taint,
+it will be force deleted from the node with zero grace period which means that Pod can be deleted immediately
+without waiting for Kubelet to responde. One important use case of this taint is node shut down case. If node
+controller or some external controller detact that node is powered off, the controller could add the following
+taint to the node.
+
+
+```
+taint := &v1.Taint{
+	Key:    nodeShutdown,
+	Value:  "true",
+	Effect: v1.TaintEffectForceEviction,
+}
+```
+
+After the node is tainted, node lifecycle controller will trigger to forcely delete all the Pods from the node.
+If Kubelet comes back healthy in the middle of force eviction process, there might be race condition between kubelet
+starting Pods while they are being deleted since force deletion does not require kubelet coordination any more. 
+To avoid race condition between controller and kubelet, kubelet also requires some additional handling when node
+is tainted which is explained in more details below.
+
+For node lifecycle controller,  we plan to add the following eviction logic:
+
+When node is marked as “NotReady” and tainted with “ForceEviction”. Taint manager will perform the following steps
+
+1. Check node condition whether Ready becomes true again. If yes, remove the taint if ForceEviction pods on the node
+is not started yet or already finished. Otherwise, skip the rest. If Ready is still false, continue with next step 2.
+2. Remove VolumeInUse list for this node and Remove VolumesAttached list
+3. Evict pods from the node with --force and 0 grace period
+
+For node Kubelet, it will first send out heartbeat by updating node status during start up.
+
+Before starting up Pod, Pod canAdmit function checks taint and pod tolerations to decide whether to admit pod or not. 
+- If node condition can not be retrieved, do not start Pods that cannot tolerant ForceEviction taint
+- If node has taint or node Ready condition is not true (heartbeat has not yet sent out), do not start Pods
```

Comment:
Are we planning to make pods backed by stateful sets or local-storage to tolerate this taint? Because evicting

Reply:
Why statefulset Pod should tolerate this taint? I think statefulset can benefit from this ForceEviction because otherwise, new Pods cannot be started on other nodes if the node is powered off.

Reply:
well usually we have volumes in pods which do have state. Usually these pods are also statefulsets. I do not see any reason why statefulset should tolerant this taint. At least I want that my database pod will move to working node if the first one is crashed. Of course there might be other problems like, database is working but kubelet is not. How we can check that?

Reply:
This taint should be used with care. The most use case is when the machine is powered off, it is safe to taint the node with ForceEviction because all processes are no longer running. For other cases, system admin/controller need to determine whether it is ok to apply this taint. Or the controller can first shutdown the node and then taint it in case of some failure.